qiskit/algorithms/gradients/param_shift_estimator_gradient.py

```diff
@@ -17,100 +17,91 @@
+
+    def _run_unique(
+        self,
+        circuits: Sequence[QuantumCircuit],
+        observables: Sequence[BaseOperator | PauliSumOp],
+        parameter_values: Sequence[Sequence[float]],
+        parameter_sets: Sequence[set[Parameter]],
         **options,
     ) -> EstimatorGradientResult:
         """Compute the estimator gradients on the given circuits."""
-        jobs, result_indices_all, coeffs_all, metadata_ = [], [], [], []
-        for circuit, observable, parameter_values_, parameters_ in zip(
-            circuits, observables, parameter_values, parameters
+        jobs, metadata = [], []
+        for circuit, observable, parameter_values_, parameter_set in zip(
+            circuits, observables, parameter_values, parameter_sets
         ):
-            # a set of parameters to be differentiated
-            if parameters_ is None:
-                param_set = set(circuit.parameters)
-            else:
-                param_set = set(parameters_)
-            metadata_.append({"parameters": [p for p in circuit.parameters if p in param_set]})
-
-            if self._gradient_circuits.get(id(circuit)):
-                gradient_circuit, base_parameter_values_all = self._gradient_circuits[id(circuit)]
-            else:
-                gradient_circuit, base_parameter_values_all = _param_shift_preprocessing(circuit)
-                self._gradient_circuits[id(circuit)] = (
-                    gradient_circuit,
-                    base_parameter_values_all,
-                )
-
-            (
-                gradient_parameter_values_plus,
-                gradient_parameter_values_minus,
-                result_indices,
-                coeffs,
-            ) = _make_param_shift_parameter_values(
-                gradient_circuit_data=gradient_circuit,
-                base_parameter_values=base_parameter_values_all,
-                parameter_values=parameter_values_,
-                param_set=param_set,
+            metadata.append({"parameters": [p for p in circuit.parameters if p in parameter_set]})
+            # Make parameter values for the parameter shift rule.
+            param_shift_parameter_values = _make_param_shift_parameter_values(
+                circuit, parameter_values_, parameter_set
             )
-            n = 2 * len(gradient_parameter_values_plus)
+            n = len(param_shift_parameter_values)
             job = self._estimator.run(
-                [gradient_circuit.gradient_circuit] * n,
+                [circuit] * n,
                 [observable] * n,
-                gradient_parameter_values_plus + gradient_parameter_values_minus,
+                param_shift_parameter_values,
                 **options,
             )
             jobs.append(job)
-            result_indices_all.append(result_indices)
-            coeffs_all.append(coeffs)
 
-        # combine the results
         try:
             results = [job.result() for job in jobs]
```

Comment:
While in principle this line enables async execution of the jobs, which should be good for performance, in practice we have found that when dealing with the runtime primitives, it is significantly faster to batch all circuits together and submit a single primitive job than sending multiple jobs. @mrossinek and I opened this draft PR (#9077) with the old gradient design. And I think it would be good to have in all of the gradient classes after this refactoring.

Reply:
Here is my suggestion to apply this batching, but maybe you can come up with a better design:

```
    def _run_unique(
        self,
        circuits: Sequence[QuantumCircuit],
        observables: Sequence[BaseOperator | PauliSumOp],
        parameter_values: Sequence[Sequence[float]],
        parameter_sets: Sequence[set[Parameter]],
        **options,
    ) -> EstimatorGradientResult:
        """Compute the estimator gradients on the given circuits."""
        jobs, metadata = [], []
        job_circuits, job_observables, job_param_values = [], [], []
        all_n = []
        for circuit, observable, parameter_values_, parameter_set in zip(
            circuits, observables, parameter_values, parameter_sets
        ):
            metadata.append({"parameters": [p for p in circuit.parameters if p in parameter_set]})
            # Make parameter values for the parameter shift rule
            param_shift_parameter_values = _make_param_shift_parameter_values(
                circuit, parameter_values_, parameter_set
            )
            n = len(param_shift_parameter_values)

            # Combine inputs to the estimator
            job_circuits.extend([circuit] * n)
            job_observables.extend([observable] * n)
            job_param_values.extend(param_shift_parameter_values)
            all_n.append(n)

        # Send single job with all circuits
        jobs = self._estimator.run(
            job_circuits,
            job_observables,
            job_param_values,
            **options,
        )
        try:
            results = jobs.result()
        except Exception as exc:
            raise AlgorithmError("Estimator job failed.") from exc

        # Compute the gradients
        gradients = []
        partial_sum_n = 0
        for n in all_n:
            result = results.values[partial_sum_n: partial_sum_n + n]
            gradient_ = (result[:n//2] - result[n//2:]) / 2
            gradients.append(gradient_)
            partial_sum_n += n

        opt = self._get_local_options(options)
        return EstimatorGradientResult(gradients=gradients, metadata=metadata, options=opt)
```

Reply:
Something that might be important to note is that by bundling together these circuits, we might be pushing the limits of how many circuits the primitives can handle, but my take here is that this is not something we should worry about, but should be handled by the actual primitives (but still good to keep in mind in case we see strange behaviours in the future).